pexpect/replwrap.py

```diff
@@ -61,11 +61,48 @@ def set_prompt(self, orig_prompt, prompt_change):
         self.child.expect(orig_prompt)
         self.child.sendline(prompt_change)
 
-    def _expect_prompt(self, timeout=-1):
+    def _expect_prompt(self, timeout=-1, async_=False):
         return self.child.expect_exact([self.prompt, self.continuation_prompt],
-                                       timeout=timeout)
+                                       timeout=timeout, async_=async_)
 
-    def run_command(self, command, timeout=-1):
+    def _run_command_async(self, cmdlines, timeout=-1):
+        import asyncio
+        loop = asyncio.get_event_loop()
+        if hasattr(loop, 'create_future'):
+            result = loop.create_future()
+        else:
+            result = asyncio.Future(loop=loop)
+        cmd_count = len(cmdlines)
+        out = []
+
+        # Called recursively through callbacks
+        def _run_line(fut=None):
+            if len(cmdlines) < cmd_count:
+                # gather output after the first line has been sent
+                out.append(self.child.before)
```

Comment:
This async recursive structure is not very easy to follow, in my opinion. Would it be better written as a coroutine?